packages/shared/lib/services/providers.ts

```diff
@@ -22,49 +26,96 @@ export function getProvider(providerName: string): Provider | null {
     return providers?.[providerName] ?? null;
 }
 
-function getProvidersPath() {
+export async function launchProvidersSync() {
+    await loadProviders();
+
+    if (providersUrl) {
+        logger.info(`monitoring providers from ${providersUrl} every ${reloadInterval}ms`);
+        setInterval(async () => {
+            try {
+                await loadProviders();
+            } catch (err) {
+                logger.error('Failed to load providers.yaml', err);
+            }
+        }, reloadInterval);
+    }
+}
+
+async function loadProviders() {
+    const maybeNewProviders = await loadProvidersRaw();
+    const newProvidersHash = createHash('sha1').update(maybeNewProviders).digest('hex');
+
+    if (newProvidersHash !== providersHash) {
+        providersHash = newProvidersHash;
+        providers = parseProviders(maybeNewProviders);
+        logger.info(`providers updated to ${providersHash}`);
+    }
+}
+
+async function getProvidersPath() {
     // find the providers.yaml file
     // recursively searching in parent directories
-    const findProvidersYaml = (dir: string): string => {
+    const findProvidersYaml = async (dir: string): Promise<string> => {
         const providersYamlPath = path.join(dir, 'providers.yaml');
-        if (fs.existsSync(providersYamlPath)) {
+
+        try {
+            await fs.stat(providersYamlPath);
             return providersYamlPath;
+        } catch {
+            const parentDir = path.dirname(dir);
+            if (parentDir === dir) {
+                throw new NangoError('providers_yaml_not_found');
+            }
+            return findProvidersYaml(parentDir);
         }
-        const parentDir = path.dirname(dir);
-        if (parentDir === dir) {
-            throw new NangoError('providers_yaml_not_found');
-        }
-        return findProvidersYaml(parentDir);
     };
     return findProvidersYaml(dirname());
 }
 
-function loadProvidersYaml(): Record<string, Provider> | undefined {
-    try {
-        const fileEntries = yaml.load(fs.readFileSync(getProvidersPath()).toString()) as Record<string, Provider | ProviderAlias>;
+async function loadProvidersRaw(): Promise<string> {
+    let rawFile: string | undefined;
+
+    if (!providersUrl) {
+        const providersPath = await getProvidersPath();
+        rawFile = (await fs.readFile(providersPath)).toString();
+    } else {
+        const url = new URL(providersUrl);
+
+        // bust caches
+        url.searchParams.set('t', Date.now().toString());
 
-        if (fileEntries == null) {
-            throw new NangoError('provider_template_loading_failed');
+        const response = await fetch(url);
+        if (response.ok) {
+            rawFile = await response.text();
+        } else {
+            throw new NangoError('provider_template_loading_fetch_failed');
         }
+    }
 
-        for (const key in fileEntries) {
-            const entry = fileEntries[key];
+    return rawFile;
+}
```

Comment:
should we back up to load it from file if it fails to fetch the url or fails to load it?